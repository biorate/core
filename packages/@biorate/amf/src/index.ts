// Internal buffer for encoding
let tempBuffer = Buffer.allocUnsafe(1024);

// Internal buffer current offset
let tempPos = 0;

// Index for string reference encoding
let encodeCounter = 0;

// Storage for string reference encoding
let encodeCache: Record<string, number> = {};

// Storage for string reference decoding
let decodeCache: string[] = [];

/**
 * @description Resize buffer twice
 * @param {number} byte
 */
function resize(byte: number) {
  tempBuffer = Buffer.concat([tempBuffer, Buffer.allocUnsafe(byte)]);
}

/**
 * @description Write data to buffer
 * @param {*} data - Data to serialize and write
 */
function write(data: any) {
  tempPos = 0;
  encodeCounter = 0;
  encodeCache = {};
  writeValue(data);
  return tempBuffer.slice(0, tempPos);
}

/**
 * @description Check type of data and write recursive
 * @param {*} data - Data to serialize and write
 */
function writeValue(data: any) {
  if (data === undefined) return writeByte(0);
  if (data === null) return writeByte(1);
  if (data === false) return writeByte(2);
  if (data === true) return writeByte(3);

  var type = typeof data;

  if (type === 'number') {
    if (data >= 0 && data < 0x10000000 && data % 1 === 0) writeInt(data);
    else writeFloat(data);
  } else if (type === 'string') {
    writeString(data);
  } else if (Array.isArray(data)) {
    writeArray(data);
  } else if (type === 'object') {
    if (data instanceof Date) writeDate(data);
    else writeObject(data);
  } else {
    throw new Error(`${type} types not supported. Please check input data.`);
  }
}

/**
 * @description Write Date to internal buffer
 * @param {Number} value - Value to write
 */
function writeDate(value: Date) {
  writeByte(8);
  writeFloat(value.getTime(), false);
}

/**
 * @description Write byte to internal buffer
 * @param {Number} value - Value to write
 */
function writeByte(value: number) {
  if (tempBuffer.length <= tempPos) resize(1);
  tempBuffer.writeUInt8(value, tempPos);
  tempPos += 1;
}

/**
 * @description Write float64 to internal buffer
 * @param {Number} value - Value to write
 * @param {Boolean} [marker=true] - False if float marker should not be written
 */
function writeFloat(value: number, marker = true) {
  if (marker !== false) writeByte(5);
  if (tempBuffer.length <= tempPos + 8) resize(8);
  tempBuffer.writeDoubleBE(value, tempPos);
  tempPos += 8;
}

/**
 * @description Write uInt29 to internal buffer
 * @param {Number} value - Value to write
 * @param {Boolean} [marker=true] - False if int marker should not be written
 */
function writeInt(value: number, marker = true) {
  if (marker !== false) writeByte(4);
  if (value < 0x80) {
    writeByte(value);
    return;
  }
  if (value < 0x4000) {
    writeByte(0x80 | ((value >> 7) & 0x7f));
    writeByte(value & 0x7f);
    return;
  }
  if (value < 0x200000) {
    writeByte(0x80 | ((value >> 14) & 0x7f));
    writeByte(0x80 | ((value >> 7) & 0x7f));
    writeByte(value & 0x7f);
    return;
  }
  writeByte(0x80 | ((value >> 22) & 0x7f));
  writeByte(0x80 | ((value >> 15) & 0x7f));
  writeByte(0x80 | ((value >> 8) & 0x7f));
  writeByte(value & 0xff);
}

/**
 * @description Write string to internal buffer
 * @param {String} value - String to write
 * @param {Boolean} [marker=true] - False if int marker should not be wrote
 */
function writeString(value: string, marker = true) {
  let index = encodeCache[value];

  if (marker !== false) writeByte(6);

  if (index !== undefined) {
    writeInt(index << 1, false);
    return;
  }

  index = Buffer.byteLength(value);
  writeInt((index << 1) + 1, false);

  if (index) {
    encodeCache[value] = encodeCounter;
    ++encodeCounter;
    if (tempBuffer.length <= tempPos + index) resize(index);
    tempBuffer.write(value, tempPos);
    tempPos += index;
  }
}

/**
 * @description Write array to internal buffer
 * @param {Array} value - Array to write
 */
function writeArray(value: unknown[]) {
  let len = value.length;

  writeByte(9);
  writeInt((len << 1) + 1, false);
  writeString('', false);

  for (let i = 0; i < len; ++i) {
    // Should we check for function type?
    // We write size of array before, and should add one more loop for check
    // It will be expensive
    writeValue(value[i]);
  }
}

/**
 * @description Write object to internal buffer
 * @param {Object} value - Object to write
 */
function writeObject(value: Record<string, unknown>) {
  writeByte(10);
  writeInt(11, false);
  writeString('', false);

  let keys = Object.keys(value);

  for (let i = 0, l = keys.length; i < l; ++i) {
    let name = keys[i];
    let item = value[name];

    // For consistency, we should remove this check
    // But objects frequently have methods
    if (typeof item === 'function')
      throw new Error(`Function types not supported. Please check input data.`);

    writeString(name, false);
    writeValue(item);
  }

  writeString('', false);
}

/**
 * @description Deserialize AMF3 byte buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function read<T = any>(buffer: Buffer): T {
  tempPos = 0;
  decodeCache.length = 0;

  return readValue(buffer);
}

/**
 * @description Check flag type and read unserialized value
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readValue(buffer: Buffer): any {
  switch (readByte(buffer)) {
    case 0:
      return undefined;
    case 1:
      return null;
    case 2:
      return false;
    case 3:
      return true;
    case 4:
      return readInt(buffer);
    case 5:
      return readFloat(buffer);
    case 6:
      return readString(buffer);
    case 8:
      return readDate(buffer);
    case 9:
      return readArray(buffer);
    case 10:
      return readObject(buffer);
  }
}

/**
 * @description Read byte from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readByte(buffer: Buffer) {
  tempPos += 1;
  return buffer.readUInt8(tempPos - 1);
}

/**
 * @description Read Date object from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readDate(buffer: Buffer) {
  tempPos += 8;
  return new Date(buffer.readDoubleBE(tempPos - 8));
}

/**
 * @description Read float64 from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readFloat(buffer: Buffer) {
  tempPos += 8;
  return buffer.readDoubleBE(tempPos - 8);
}

/**
 * @description Read int29 from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readInt(buffer: Buffer) {
  let value: number;
  let byte: number;

  byte = readByte(buffer);
  if (byte < 0x80) return byte;

  value = (byte & 0x7f) << 7;
  byte = readByte(buffer);
  if (byte < 0x80) return value | byte;

  value = (value | (byte & 0x7f)) << 7;
  byte = readByte(buffer);
  if (byte < 0x80) return value | byte;

  value = (value | (byte & 0x7f)) << 8;
  byte = readByte(buffer);
  return value | byte;
}

/**
 * @description Read string from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readString(buffer: Buffer) {
  let len = readInt(buffer);

  if ((len & 1) === 0) return decodeCache[len >> 1];

  len = len >> 1;

  let string = '';

  if (len) {
    string = buffer.toString('utf8', tempPos, tempPos + len);
    decodeCache.push(string);
    tempPos += len;
  }

  return string;
}

/**
 * @description Read array from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readArray(buffer: Buffer) {
  let len = readInt(buffer);
  len = len >> 1;

  readString(buffer);

  let result = [];

  for (let i = 0; i < len; ++i) result.push(readValue(buffer));

  return result;
}

/**
 * @description Read object from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readObject(buffer: Buffer) {
  let name;
  readInt(buffer);
  readString(buffer);

  let result: Record<string, unknown> = {};

  while ((name = readString(buffer))) result[name] = readValue(buffer);

  return result;
}

/**
 * @description AMF3 encode function
 *
 * Links:
 * - [wikipedia](https://en.wikipedia.org/wiki/Action_Message_Format)
 * - [specification] https://www.adobe.com/content/dam/acom/en/devnet/pdf/amf-file-format-spec.pdf
 *
 * Limitations:
 * - Function is not supported
 * - XML is not supported
 * - ByteArray is not supported
 * - Dictionary is not supported
 * - Vectors is not supported
 *
 * @example
 * ```ts
 * import { encode } from '@biorate/amf';
 *
 * const buffer = encode({ test: 1 });
 *
 * console.log(buffer); // <Buffer 0a 0b 01 09 74 65 73 74 04 01 01>
 * ```
 */
export const encode = write;

/**
 * @description AMF3 decode function
 *
 * Limitations:
 *   - see encode description section
 *
 * @example
 * ```ts
 * import { encode, decode } from '@biorate/amf';
 *
 * const buffer = encode({ test: 1 });
 *
 * console.log(decode(buffer)); // { test: 1 }
 * ```
 */
export const decode = read;
