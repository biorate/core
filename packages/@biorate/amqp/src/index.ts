import { EventEmitter } from 'events';
import { injectable } from '@biorate/inversion';
import { Connector } from '@biorate/connector';
import { events } from '@biorate/tools';
import { connect, ChannelWrapper } from 'amqp-connection-manager';
import { AmqpCantConnectError, ChannelNotExistsError } from './errors';
import { IAmqpConfig, IAmqpConnection, ICreateChannelOpts } from './interfaces';

export { ConsumeMessage } from 'amqplib/properties';
export * from 'amqp-connection-manager';
export * from './errors';
export * from './interfaces';

/**
 * @description Amqp connector
 *
 * ### Features:
 * - connection manager for AMQP
 *
 * @example
 * ```
 * import { inject, container, Types, Core } from '@biorate/inversion';
 * import { IConfig, Config } from '@biorate/config';
 * import { AmqpConnector } from '@biorate/amqp';
 *
 * const connectionName = 'amqp';
 * const channelName = 'test';
 *
 * export class Root extends Core() {
 *   @inject(AmqpConnector) public connector: AmqpConnector;
 * }
 *
 * container.bind<IConfig>(Types.Config).to(Config).inSingletonScope();
 * container.bind<AmqpConnector>(AmqpConnector).toSelf().inSingletonScope();
 * container.bind<Root>(Root).toSelf().inSingletonScope();
 *
 * container.get<IConfig>(Types.Config).merge({
 *   Amqp: [
 *     {
 *       name: connectionName,
 *       urls: ['amqp://localhost:5672'],
 *     },
 *   ],
 * });
 *
 * const root = <Root>container.get<Root>(Root);
 *
 * (async () => {
 *   root.connector.createChannel(connectionName, {
 *     name: channelName,
 *     json: true,
 *     setup: async (channel: Channel) => {
 *       await channel.assertExchange('test-exchange', 'topic');
 *       await channel.assertQueue('test-queue', { exclusive: true, autoDelete: true });
 *       await channel.bindQueue('test-queue', 'test-exchange', '#send');
 *       await channel.consume('test-queue', (data: ConsumeMessage | null) => {
 *         console.log(data?.content?.toString?.()); // {"test": 1}
 *       });
 *     },
 *   });
 *   root.connector.channel(channelName)!.publish('test-exchange', '#send', { test: 1 });
 * })();
 * ```
 */
@injectable()
export class AmqpConnector extends Connector<IAmqpConfig, IAmqpConnection> {
  /**
   * @description Private connections storage
   */
  private '#connections': Map<string, IAmqpConnection>;
  /**
   * @description Private link to selected (used) connection
   */
  private '#current': IAmqpConnection | undefined;
  /**
   * @description Namespace path for fetching configuration
   */
  protected readonly namespace = 'Amqp';
  /**
   * @description Channels storage
   */
  protected readonly channels = new Map<string, ChannelWrapper>();
  /**
   * @description Create connection
   */
  protected async connect(config: IAmqpConfig) {
    let connection: IAmqpConnection;
    try {
      connection = connect(config.urls, config.options);
      await events.once(<EventEmitter>(<unknown>connection), 'connect');
      connection.on('disconnect', console.warn);
    } catch (e: unknown) {
      throw new AmqpCantConnectError(<Error>e);
    }
    return connection;
  }
  /**
   * @description Create channel method
   */
  public createChannel(name: string, options: ICreateChannelOpts) {
    const channel = this.connection(name).createChannel(options);
    this.channels.set(options.name, channel);
    return channel;
  }
  /**
   * @description Get channel method
   */
  public channel(name: string) {
    if (!this.channels.has(name)) throw new ChannelNotExistsError(name);
    return this.channels.get(name);
  }
}
