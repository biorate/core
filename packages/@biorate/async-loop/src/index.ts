import { timer } from '@biorate/tools';

/**
 * @description Async loop implementation
 *
 * @example
 * ```
 * import { AsyncLoop } from '@biorate/async-loop';
 *
 * (async () => {
 *   let i = 0;
 *   const loop = new AsyncLoop(
 *     () => {
 *       console.log(++i); // 1, 2, 3 ... e.t.c
 *     },
 *     console.error,
 *     1000,
 *   );
 * })();
 * ```
 */
export class AsyncLoop {
  #process: () => Promise<void> | void;

  #error: (e: unknown) => Promise<void> | void;

  #interval: number;

  #paused = false;

  #stoped = false;

  public constructor(
    process: () => Promise<void> | void,
    error?: () => Promise<void> | void,
    interval = 1000,
  ) {
    this.#process = process;
    this.#error = error ?? console.error;
    this.#interval = interval;
    this.process().catch(this.#error);
  }

  public pause() {
    this.#paused = true;
  }

  public resume() {
    this.#paused = false;
  }

  public stop() {
    this.#stoped = true;
  }

  protected async process() {
    while (true) {
      await timer.wait(this.#interval);
      if (this.#paused) continue;
      if (this.#stoped) break;
      try {
        await this.#process();
      } catch (e: unknown) {
        await this.#error(e);
      }
    }
  }
}
