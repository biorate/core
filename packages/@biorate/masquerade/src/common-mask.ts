import { inject, Types } from '@biorate/inversion';
import { IConfig } from '@biorate/config';
import type { IMask, IMaskOptions } from '../interfaces';

export abstract class CommonMask implements IMask {
  @inject(Types.Config) protected config: IConfig;

  protected abstract options?: IMaskOptions | null;

  protected get name() {
    return this.constructor.name;
  }

  protected get maskChar() {
    return this.options?.maskChar ?? '*';
  }

  protected get enabled() {
    return this.options?.enabled ?? true;
  }

  public process(text: string, options?: IMaskOptions) {
    if (!this.enabled) return text;
    this.options = options;
    return this.parse(text);
  }

  protected abstract parse(text: string): string;
}
