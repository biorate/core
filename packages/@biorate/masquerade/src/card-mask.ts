import { CommonMask } from './common-mask';
import type { ICardMaskOptions } from '../interfaces';

export class CardMask extends CommonMask {
  protected options?: ICardMaskOptions;

  protected get firstDigits() {
    return this.options?.firstDigits ?? 6;
  }

  protected get lastDigits() {
    return this.options?.lastDigits ?? 4;
  }

  protected parse(text: string) {
    return text.replace(/\b(?:\d[ \t\-\.]*?){15,18}\d\b/g, (card) => {
      const digits = card.replace(/\D/g, '');
      const len = digits.length;
      // Check the length (standard cards: 13-19 digits)
      if (len < 13 || len > 19) return card;
      let masked = '';
      let digitCount = 0;
      for (const char of card) {
        if (/\d/.test(char)) {
          digitCount++;
          // Save the first 6 and last 4 digits
          if (digitCount <= this.firstDigits || digitCount > len - this.lastDigits) {
            masked += char;
          } else {
            masked += this.maskChar;
          }
        } else {
          masked += char;
        }
      }
      return masked;
    });
  }
}
