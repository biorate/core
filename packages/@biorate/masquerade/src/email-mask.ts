import { CommonMask } from './common-mask';
import type { IEmailMaskOptions } from '../interfaces';

export class EmailMask extends CommonMask {
  protected options?: IEmailMaskOptions;

  protected get unmaskedEndChars() {
    return this.options?.unmaskedEndChars ?? 1;
  }

  protected get unmaskedStartChars() {
    return this.options?.unmaskedStartChars ?? 1;
  }

  protected parse(text: string) {
    return text.replace(/\b[\w.%+-]+@[\w.-]+\.[a-z]{2,}\b/gi, (email) => {
      const [local, domain] = email.split('@');
      if (local.length <= 1) return email;
      return (
        local.slice(0, this.unmaskedStartChars) +
        this.maskChar.repeat(local.length - this.unmaskedStartChars) +
        '@' +
        this.maskChar.repeat(domain.length - this.unmaskedEndChars) +
        domain.slice(-this.unmaskedEndChars)
      );
    });
  }
}
