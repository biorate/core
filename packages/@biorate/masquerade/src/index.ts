import { merge } from 'lodash';
import { JsonMask2Configs, maskJSON2 } from 'maskdata';
import { Types, injectable, inject, init } from '@biorate/inversion';
import { IConfig } from '@biorate/config';
import { CommonMask } from './common-mask';
import { MaskConstructor, IMaskOptions, IMasqueradeConfig } from '../interfaces';

export * from './common-mask';
export * from './email-mask';
export * from './phone-mask';
export * from './card-mask';
/**
 * @description
 * Masquerade provides functionality to redact sensitive info
 *
 * @example
 * ```
 * import { Masquerade } from '@biorate/masquerade';
 *
 * Masquerade.configure({ maskJSON2: { emailFields: ['email'] } });
 *
 * const result = Masquerade.processJSON({ email: 'test@email.com' });
 *
 * console.log(result); // { "email": "tes*@*******om" }
 * ```
 *
 * ```
 * import { Masquerade } from '@biorate/masquerade';
 *
 * Masquerade.use(EmailMask).use(PhoneMask).use(CardMask);
 *
 * const result = Masquerade.processString(
 *   `user@example.com, +79231231224, 4111 1111 1111 1111 (Visa), 5500-0000-0000-0004 (MC)`
 * );
 *
 * console.log(result); // u***@**********m, +*******1224, **** **** **** 1111 (Visa), ****-****-****-0004 (MC)
 * ```
 */
@injectable()
export class Masquerade {
  protected static config: IMasqueradeConfig = {};

  protected static maskers = new Map<string, CommonMask>();

  public static configure(config: IMasqueradeConfig | null) {
    if (!config) return this;
    merge(this.config, config);
    return this;
  }

  public static get maskdataEnabled() {
    return !!Masquerade.config?.maskJSON2;
  }

  public static use<T extends CommonMask>(Mask: MaskConstructor<T>) {
    this.maskers.set(Mask.name, new Mask());
    return this;
  }

  public static unuse<T extends CommonMask>(Mask: MaskConstructor<T>) {
    this.maskers.delete(Mask.name);
    return this;
  }

  public static processJSON<T extends object>(data: T, options?: JsonMask2Configs) {
    const opts = options ?? Masquerade.config?.maskJSON2;
    if (!opts) return data;
    return maskJSON2(data, opts);
  }

  public static processString(text: string, options?: IMaskOptions) {
    for (const [, mask] of Masquerade.maskers) {
      const opts =
        options ??
        <IMaskOptions>(
          (<unknown>Masquerade.config?.[<keyof IMasqueradeConfig>mask.constructor.name])
        );
      text = mask.process(text, opts);
    }
    return text;
  }

  @inject(Types.Config) protected config: IConfig;

  @init() protected initialize() {
    Masquerade.configure(this.config.get<IMasqueradeConfig | null>('Masquerade', null));
  }
}
