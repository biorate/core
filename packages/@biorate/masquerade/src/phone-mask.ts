import { CommonMask } from './common-mask';
import type { IPhoneMaskOptions } from '../interfaces';

export class PhoneMask extends CommonMask {
  protected options?: IPhoneMaskOptions;

  protected regexp = /(?:\+7|8)[\s\-()]*\d{3}[\s\-()]*\d{3}[\s\-()]*\d{2}[\s\-()]*\d{2}/g;

  protected get minDigits() {
    return this.options?.minDigits ?? 7;
  }

  protected get preserveCount() {
    return this.options?.preserveCount ?? 4;
  }

  protected parse(text: string) {
    return text.replace(this.regexp, (phone) => {
      const digits = phone.match(/\d/g);
      if (!digits || digits.length < this.minDigits) return phone;
      let masked = '';
      let digitCount = 0;
      for (const char of phone) {
        if (/\d/.test(char)) {
          masked +=
            digitCount++ < digits.length - this.preserveCount ? this.maskChar : char;
        } else {
          masked += char;
        }
      }
      return masked;
    });
  }
}
