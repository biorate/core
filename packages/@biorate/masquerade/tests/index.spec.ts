import { expect, use } from 'chai';
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot';
import { text } from './__mocks__';
import { Masquerade } from '../src';

use(jestSnapshotPlugin());

describe('@biorate/masquerade', function () {
  it('processJSON', () => {
    expect(
      Masquerade.processJSON({ email: 'test@email.com' }, { emailFields: ['email'] }),
    ).toMatchSnapshot();
  });

  it('processString', () => {
    expect(Masquerade.processString(text)).toMatchSnapshot();
  });
});
