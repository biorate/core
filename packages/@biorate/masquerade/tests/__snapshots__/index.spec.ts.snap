// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@biorate/masquerade processJSON 1`] = `
Object {
  "email": "tes*@*******om",
}
`;

exports[`@biorate/masquerade processString 1`] = `
"2025-08-06T09:00:06.382Z

2025-08-06 09:00:06

О компании \\"ФинансГарант\\"
Приветствуем вас! Мы — команда профессионалов, посвятивших себя защите ваших финансовых интересов. Наши эксперты всегда готовы помочь:

Наша команда
Александр Петров — руководитель отдела безопасности:
Контакты: a*******@**************u, рабочий телефон: +* (***) ***-43-21
Личный WhatsApp: *-***-***-4567

Мария Иванова-Сидорова — финансовый аналитик:
Email: m********@***************m, Tel: 495 987 65 43
Экстренная связь: 916-765-4321

John Smith — международные партнерства:
j******@***************g, Mobile: +44 20 7123 4567
Для платежей: Карта Visa 4111 11** **** 1111 (до 08/2025)

Контактная информация
Центральный офис:
125167, Москва, Ленинградский проспект, д. 39, стр. 80
Телефоны:

Общий: * (***) ***-14-14

Бухгалтерия: +* *** ***-45-67 доб. 506

Факс: 495-123-45-68

Электронная почта:

Общие вопросы: i***@**************u

Техподдержка: s******@***************m

Партнерство: p*******@*************g

Платежные реквизиты
Для международных переводов:

Банковская карта: Mastercard 5500 00** **** 0004 (Иванов И.И.)

Счет в EUR: DE91 1000 00** **** **67 89

Криптокошелек: 0x742d35Cc6634C0532925a3b844Bc454e4438f44e

Истории успеха
Наш клиент Сергей Кузнецов (s**********@********m, тел. 903-123-4567) увеличил капитал на 45% за год. Его отзыв:
\\"Благодаря советам Марии Ивановой (m********@***************m) и использованию карты Amex 3782 822463 10005 мои инвестиции стали работать эффективнее!\\"

Технические контакты
При обнаружении уязвимостей обращайтесь:

Сергей Васильев — Head of Security
PGP: s*******@**************u
Телефон: +* *** ***-11-22
Для вознаграждений: Visa 4242 42** **** 4242

Примечание безопасности:
Все контакты защищены системой маскировки данных. Реальные данные доступны только авторизованным сотрудникам через защищенные каналы. Для проверки сервиса используйте тестовые платежные реквизиты:

Карта: 4000 00** **** 3220 (Visa Test)

Телефон: +* (***) ***-00-01

Содержание персональных данных:
Имена (8 вхождений):

Александр Петров

Мария Иванова-Сидорова

John Smith

Иванов И.И.

Сергей Кузнецов

Сергей Васильев

m.ivanova (имя в email)

s.kuznetsov (имя в email)

Телефоны (8 номеров):

+* (***) ***-43-21

*-***-***-4567

495 987 65 43

916-765-4321

+44 20 7123 4567

* (***) ***-14-14

+* *** ***-45-67 доб. 506

+* *** ***-11-22

Email (7 адресов):

a*******@**************u

m********@***************m

j******@***************g

i***@**************u

s******@***************m

p*******@*************g

s**********@********m

Банковские карты (5 номеров):

4111 11** **** 1111

5500 00** **** 0004

3782 822463 10005

4242 42** **** 4242

4000 00** **** 3220

Другие данные:

Банковский счет: DE91 1000 00** **** **67 89

Криптокошелек: 0x742d35Cc6634C0532925a3b844Bc454e4438f44e

Почтовый адрес: 125167, Москва, Ленинградский проспект, д. 39, стр. 80

Этот текст содержит множественные вхождения всех требуемых типов персональных данных в различных форматах, что идеально подходит для тестирования систем маскировки.
"
`;
